app/src/org/commcare/dalvik/activities/EntityMapActivity.java

```diff
@@ -1,288 +1,186 @@
-            int lat = (int) (location.getLatitude() * 1E6);
-            int lng = (int) (location.getLongitude() * 1E6);
-            GeoPoint point = new GeoPoint(lat, lng);
-            map.getController().animateTo(point);
-        }
-        
-        map.getOverlays().add(mMyLocationOverlay);
-        //The overlay crashes out if you try to draw it and it's empty,
-        //so only add it if we found something.
-        if(mEntityOverlay.size() > 0) {
-            map.getOverlays().add(mEntityOverlay);
-        }
-        map.getController().setZoom(18);
-        map.setClickable(true);
-        map.setEnabled(true);
-        Log.d(TAG, "Done loading");
+        Log.d(TAG, "Loaded. " + entityLocations.size() +" addresses discovered, " + bogusAddresses + " could not be located");
     }
-    
-    private String getLicenseKey() {
-        //If there's a defined debug key in the local environment, use that.
-        int debugId = this.getResources().getIdentifier("maps_api_key_debug","string", this.getPackageName());
-        if(debugId == 0) { 
-            return BuildConfig.MAPS_API_KEY;
+
+    private LatLng getLatLngFromAddress(String address) {
+        LatLng location = null;
+        try {
+            GeoPointData data = new GeoPointData().cast(new UncastData(address));
+            if(data != null) {
+                location = new LatLng(data.getLatitude(), data.getLongitude());
+            }
+        } catch(Exception ex) {
         }
-        return this.getString(debugId);
-    }
 
-    private EvaluationContext entityContext;
+        // Geocaching code removed
 
-    private EvaluationContext getEC() {
-        if(entityContext == null) {
-            entityContext = session.getEvaluationContext(getInstanceInit());
-        }
-        return entityContext;
-    }
-    
-    private AndroidInstanceInitializer getInstanceInit() {
-        return new AndroidInstanceInitializer(session);
+        return location;
     }
 
     @Override
-    protected void onStop() {
-        super.onStop();
-        mMyLocationOverlay.disableCompass(); 
-        mMyLocationOverlay.disableMyLocation();
+    public void onMapReady(final GoogleMap map) {
+        mMap = map;
+
+        // Find bounding region of markers
+        LatLngBounds.Builder builder = new LatLngBounds.Builder();
+        for (Pair<Entity<TreeReference>, LatLng> entityLocation: entityLocations) {
+            Marker marker = map.addMarker(new MarkerOptions()
+                    .position(entityLocation.second)
+                    .title(entityLocation.first.getFieldString(0))
+                    .snippet(entityLocation.first.getFieldString(1)));
+            markerReferences.put(marker, entityLocation.first.getElement());
+            builder.include(entityLocation.second);
+        }
+        final LatLngBounds bounds = builder.build();
+
+        // Move camera to be include all markers
+        // TODO: does this work for 0 or 1 marker?

```

Comment:
Maybe you can listen for the bounds changing via http://stackoverflow.com/a/13942063 and then store it such that if you navigate away from the activity and come back it can use those bounds.  I think you can store that info using `onSaveInstanceState`.

Related, what happens when you rotate the device while viewing the map? Does it recenter to the original bounds? If so, this will also fix that